File: 107_binary_tree_level_order_traversal_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // Time Complexity: O(num(nodes))
    // Space Complexity: O(num(nodes))
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>>  result = new ArrayList();
        if(root==null) return result;
        List<Integer> temp = new ArrayList();
        Queue<TreeNode> q = new LinkedList();
        q.add(root);
        while(!q.isEmpty()){
            temp = new ArrayList();
            int size = q.size();
            for(int i = 0; i < size; i++){
                TreeNode node = q.remove();
                temp.add(node.val);
                if(node.left != null) q.add(node.left);
                if(node.right != null) q.add(node.right);
            }
            result.add(temp);
        }

        Collections.reverse(result);
        return result;
    }
}

File: python/107_binary_tree_level_order_traversal_II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        queue = deque()
        if root is not None:
            queue.append(root)
            queue.append(None)

        res = []
        row = []
        while len(queue) > 0:
            node = queue.popleft()
            if node is not None:
                row.append(node.val)
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
            else:
                res.append(list(row))
                row = []
                if len(queue) > 0:
                    queue.append(None)
        res.reverse()
        return res

